Supports flexible input partition specs.

markblee opened the pull request.

ruomingp reviewed (approved):
Did you validate that training still works?

markblee commented:
> Did you validate that training still works?

Added some links in the internal description. This should be a no-op to internal runs but I fixed and added a test for fuji.

ruomingp commented on axlearn/common/input_base.py:
```diff
@@ -2,23 +2,121 @@
+
+
+def partition_by_path_rank(
+    path_rank_to_partition: dict[PathAndRank, PartitionSpec],
```
Nit: maybe we don't really need a dict:

```suggestion
    path_rank_to_partition: Sequence[IntputFieldPartitionSpec],
```

it simplifies the data structure and `IntputFieldPartitionSpec` is more extensible than `PartitionSpec`.

markblee replied:
Thanks -- to clarify, did you mean to specify `InputFieldPartitionSpec` as a `NamedTuple` of `(path, rank, PartitionSpec)`, or something else?

(To unblock my next PR, I am happy to address this in a follow-up, if that's OK.)

ruomingp replied:
> Thanks -- to clarify, did you mean to specify `InputFieldPartitionSpec` as a `NamedTuple` of `(path, rank, PartitionSpec)`

Yes, a namedtuple, a dataclass, or a chex.struct class.

>, or something else?
> 
> (To unblock my next PR, I am happy to address this in a follow-up, if that's OK.)

Sure.
